udpTester.py

```diff
@@ -57,206 +71,204 @@ def close(self):
     # to have reasonable response time when pressing ctrl c to stop the application.
     def wait(self, timeout):
         miniTimeout = 0.3
-        max_attempts = math.ceil( timeout / miniTimeout )
+        max_attempts = math.ceil(timeout / miniTimeout)
         for attempt in range(0, max_attempts):
-            events = self.sel.select(miniTimeout)
-            if( events != [] ):
+            if self.sel.select(miniTimeout):
                 return True
         raise TimeoutError("socketWaitset.wait() timed out")
 
-class udpMetricsReportItem:
-    percentile = 0
-    sampleCount = 0
-    minimum = 0
-    average = 0
-    maximum = 0
-    deviation = 0
-    def print(self, prefix):
-        print("{}{:5.1f} % : cnt= {}, min= {:.0f}, avg= {:.0f}, max= {:.0f}, dev= {:.2f}".format(
-            prefix,
-            self.percentile,
-            self.sampleCount,
-            self.minimum,
-            self.average,
-            self.maximum,
-            self.deviation))
 
+@dataclasses.dataclass
+class udpMetricsReportItem:
+    percentile: float = 0.0
+    sampleCount: int = 0
+    totalSampleCount: int = 0
+    minimum: float = 0
+    average: float = 0
+    maximum: float = 0
+    deviation: float = 0
+
+    def __str__(self):
+        return (
+            f"{self.percentile:5.1f} % : cnt= {self.sampleCount}/{self.totalSampleCount}, min= {self.minimum:.0f},"
+            f" avg= {self.average:.0f}, max= {self.maximum:.0f}, dev= {self.deviation:.2f}"
+        )
+
+
+@dataclasses.dataclass
 class udpMetrics:
-    sampleCount = 0
-    maxCount = 0
-    values = []
-    def __init__(self, maxNofSamples):
-        self.reset()
-        self.maxCount = maxNofSamples
-
-    def reset(self):
-        self.sampleCount = 0
-        self.values = []
-
-    def addValue(self, value):
-        assert(self.sampleCount < self.maxCount)
-        if(self.sampleCount < self.maxCount):
-            self.values.append(value)
-            self.sampleCount += 1
-
-    def calculateReport(self, percentile):
-        if( self.sampleCount <= 0 ):
+    max_number_of_samples: int
+    samples: list = dataclasses.field(default_factory=list)
```

Comment:
The udpMetrics class does not deal with the samples themselves, but with measured latency values. Because there is one measurement per sample, the count happens to be the same. I have noticed the latency calculated as current time - timestamp can be large or negative if the transmitter and receiver are on different machines, probably because the clocks aren't exactly the same.